lib/image-editor/class-image-editor.php

```diff
@@ -0,0 +1,260 @@
+
+		$image_path = get_attached_file( $media_id );
+
+		if ( empty( $image_path ) ) {
+			return new WP_Error( 'rest_cannot_find_attached_file', __( 'Unable to find original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		$image_editor = wp_get_image_editor( $image_path );
+		if ( ! $image_editor->load() ) {
+			return new WP_Error( 'rest_cannot_load_editor', __( 'Unable to load original media file.', 'gutenberg' ), array( 'status' => 500 ) );
+		}
+
+		return array(
+			'editor' => $image_editor,
+			'path'   => $image_path,
+		);
+	}
+
+	/**
+	 * Gets the JSON response object for an image.
+	 *
+	 * @param integer $id Image ID.
+	 * @return array Image JSON.
+	 */
+	private function get_image_as_json( $id ) {
+		$path     = '/wp/v2/media/' . $id;
+		$response = rest_do_request( $path );
+
+		if ( ! $response->is_error() ) {
+			$response->set_status( 201 );
+			$response->header( 'Location', rest_url( $path ) );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Checks for the existence of an image and if it exists, return the image.
+	 *
+	 * @param array  $attachment Attachment with url to look up.
+	 * @param string $target_file Target file name to look up.
+	 * @return array|false Image JSON if exists, otherwise false.
+	 */
+	private function get_existing_image( $attachment, $target_file ) {
+		$url = str_replace( basename( $attachment['url'] ), $target_file, $attachment['url'] );
+
+		$new_id = attachment_url_to_postid( $url );
+		if ( $new_id > 0 ) {
+			return $this->get_image_as_json( $new_id );
+		}
+
+		return false;
+	}
+
+	/**
+	 * Saves an edited image.
+	 *
+	 * @param array  $image_edit Image path and editor to save.
+	 * @param string $target_name Target file name to save as.
+	 * @param array  $attachment Attachment with metadata to apply.
+	 * @return array|WP_Error Image JSON if successful, WP_Error otherwise
+	 */
+	private function save_image( $image_edit, $target_name, $attachment ) {
+		$filename = rtrim( dirname( $image_edit['path'] ), '/' ) . '/' . $target_name;
+
+		// Save to disk.
+		$saved = $image_edit['editor']->save( $filename );
+
+		if ( is_wp_error( $saved ) ) {
+			return $saved;
+		}
+
+		// Update attachment details.
+		$attachment_post = array(
+			'guid'           => $saved['path'],
+			'post_mime_type' => $saved['mime-type'],
+			'post_title'     => pathinfo( $target_name, PATHINFO_FILENAME ),
+			'post_content'   => '',
+			'post_status'    => 'inherit',
+		);
+
+		// Add this as an attachment.
+		$attachment_id = wp_insert_attachment( $attachment_post, $saved['path'], 0, true );
+
+		if ( is_wp_error( $attachment_id ) ) {
+			if ( 'db_update_error' === $attachment_id->get_error_code() ) {
+				$attachment_id->add_data( array( 'status' => 500 ) );
+			} else {
+				$attachment_id->add_data( array( 'status' => 400 ) );
+			}
+
+			return $attachment_id;
+		}
+
+		// Generate thumbnails.
+		$metadata = wp_generate_attachment_metadata( $attachment_id, $saved['path'] );
```

Comment:
Why are you generating new thumbnails here? At this rate, may it better to just create a new attachment and copy over existing data to it.